### PR TITLE
Extend merge-expressions to work on Optional

### PR DESCRIPTION
**Pull request by sjakobi**
This is an exploratory implementation of the `Optional` part of https://github.com/dhall-lang/dhall-lang/issues/463.

---

TODOs:

* [x] Update `normalizeWithM`
* [x] Update `isNormalized` (if necessary)
* [x] Update type errors

### CODE REVIEW

**Comment by sjakobi** [comment]
Does this look sensible so far? In that case I'd try to update the standard too.

**Comment by sjakobi** [comment]
Example usage:

```
⊢ merge { None = 1, Some = \(n : Bool) -> if n then 2 else 3 } (None Bool)

1

⊢ merge { None = 1, Some = \(n : Bool) -> if n then 2 else 3 } (Some True)

2

⊢ merge { None = 1, Some = \(n : Bool) -> if n then 2 else 3 } (Some False)

3
```

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, this looks good to me so far

**Comment by sjakobi** [comment]
PR to change the standard: https://github.com/dhall-lang/dhall-lang/pull/860.

File: dhall/src/Dhall/Normalize.hs [hunk 1]
```diff
@@ -920,6 +920,14 @@ normalizeWithM ctx e0 = loop (Syntax.denote e0)
                                     Nothing -> Merge x' y' <$> t'
                             _ ->
                                 Merge x' y' <$> t'
+                    Some a ->
+                        case Dhall.Map.lookup "Some" kvsX of
+                            Just vX -> loop (App vX a)
+                            Nothing -> Merge x' y' <$> t'
```
Comment (sjakobi) [inline dhall/src/Dhall/Normalize.hs]: I was a bit surprised of this pattern of simply ignoring ill-typed code. Then again I'm not sure who's using `normalizeWithM` at all?!

Comment (Gabriel439) [reply dhall/src/Dhall/Normalize.hs]: @sjakobi: It's been that way since the very beginning

**Comment by sjakobi** [comment]
There are still a bunch of type errors using the bit

> You can ❰merge❱ the alternatives of a union using a record with one handler per alternative

Should I change all of these to

> You can ❰merge❱ the alternatives of a union or an ❰Optional❱using a record with one handler per alternative

?

The "alternatives" term is also a bit imprecise.

What do you think @Gabriel439?

Examples:

https://github.com/dhall-lang/dhall-haskell/blob/01605c7865d975ec3347b7336816f680747d72fc/dhall/src/Dhall/TypeCheck.hs#L2973-L2979

https://github.com/dhall-lang/dhall-haskell/blob/01605c7865d975ec3347b7336816f680747d72fc/dhall/src/Dhall/TypeCheck.hs#L3353-L3362

**Comment by Gabriel439** [comment]
@sjakobi: Yes, the error messages for `merge` should reference `Optional`

Regarding terminology, the convention I've been using is:

```dhall
< A | B >  -- Union

A  -- Alternative

< A | B >.A  -- Constructor
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -3041,7 +3046,17 @@ prettyTypeMessage (MustMergeUnion expr0 expr1) = ErrorMessages {..}
...
         \                                                                                \n\
         \                                                                                \n\
-        \... but the second argument to ❰merge❱ must be a union and not some other type. \n\
+        \... or this:                                                                    \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌─────────────────────────────────────────────────────────┐                 \n\
+        \    │ let optional = None Bool                                │                 \n\
+        \    │ let handlers = { None = False, Some = λ(x : Bool) → x } │                 \n\
+        \    │ in  merge handlers optional : Bool                      │                 \n\
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]: Is the type annotation required?

```suggestion
        \    │ in  merge handlers optional                             │                 \n\
```

Comment (sjakobi) [reply dhall/src/Dhall/TypeCheck.hs]: I have removed the type annotation in some other error messages for `merge` too.

**Pull request merged**